pkg/util/log/line_buffer.go

```diff
@@ -0,0 +1,162 @@
+type LineBufferedLogger struct {
+	buf     *threadsafeBuffer
+	entries atomic.Uint32
+	cap     uint32
+	w       io.Writer
+
+	onFlush func(entries uint32)
+}
+
+// Size returns the number of entries in the buffer.
+func (l *LineBufferedLogger) Size() uint32 {
+	return l.entries.Load()
+}
+
+// Write writes the given bytes to the line buffer, and increments the entries counter.
+// If the buffer is full (entries == cap), it will be flushed, and the entries counter reset.
+func (l *LineBufferedLogger) Write(p []byte) (n int, err error) {
+	// when we've filled the buffer, flush it
+	if l.Size() >= l.cap {
+		// Flush resets the size to 0
+		if err := l.Flush(); err != nil {
+			l.buf.Reset()
+			return 0, err
+		}
+	}
+
+	l.entries.Inc()
+
+	return l.buf.Write(p)
+}
+
+// Flush forces the buffer to be written to the underlying writer.
+func (l *LineBufferedLogger) Flush() error {
+	sz := l.Size()
+	if sz <= 0 {
+		return nil
+	}
+
+	// reset the counter
+	l.entries.Store(0)
+
+	// WriteTo() calls Reset() on the underlying buffer, so it's not needed here
+	_, err := l.buf.WriteTo(l.w)
+
+	// only call OnFlush callback if write was successful
+	if err == nil && l.onFlush != nil {
+		l.onFlush(sz)
+	}
+
+	return err
+}
+
+type LineBufferedLoggerOption func(*LineBufferedLogger)
+
+// WithFlushPeriod creates a new LineBufferedLoggerOption that sets the flush period for the LineBufferedLogger.
+func WithFlushPeriod(d time.Duration) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		go func() {
+			tick := time.NewTicker(d)
+			defer tick.Stop()
+
+			for range tick.C {
+				l.Flush()
+			}
+		}()
+	}
+}
+
+// WithFlushCallback allows for a callback function to be executed when Flush() is called.
+// The length of the buffer at the time of the Flush() will be passed to the function.
+func WithFlushCallback(fn func(entries uint32)) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		l.onFlush = fn
+	}
+}
+
+// WithPrellocatedBuffer preallocates a buffer to reduce GC cycles and slice resizing.
+func WithPrellocatedBuffer(size uint32) LineBufferedLoggerOption {
+	return func(l *LineBufferedLogger) {
+		l.buf = newThreadsafeBuffer(bytes.NewBuffer(make([]byte, 0, size)))
+	}
+}
+
+// NewLineBufferedLogger creates a new LineBufferedLogger with a configured capacity.
+// Lines are flushed when the context is done, the buffer is full, or the flush period is reached.
+func NewLineBufferedLogger(w io.Writer, cap uint32, opts ...LineBufferedLoggerOption) *LineBufferedLogger {
+	l := &LineBufferedLogger{
+		w:   w,
+		buf: newThreadsafeBuffer(bytes.NewBuffer([]byte{})),
+		cap: cap,
+	}
+
+	for _, opt := range opts {
+		opt(l)
+	}
+
+	return l
+}
+
+// threadsafeBuffer wraps the non-threadsafe bytes.Buffer.
+type threadsafeBuffer struct {
+	sync.RWMutex
+
+	buf *bytes.Buffer
+}
+
+// Read returns the contents of the buffer.
+func (t *threadsafeBuffer) Read(p []byte) (n int, err error) {
+	t.RLock()
+	defer t.RUnlock()
```

Comment:
I think this shouldn't be a `RWMutex` but a `Mutex` instead. I think we're never reading concurrently here so there's no advantage on that, but locking a `sync.RWMutex` has a higher cost than a `sync.Mutex`

Reply:
I implemented this a while ago in Loki, so I can't quite recall why I did this, so you may be correct.

Reply:
I tested with a `sync.Mutex` (results in `/tmp/after`).

```bash
$ benchstat /tmp/before /tmp/after                                                       
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/util/log
                              │ /tmp/before  │             /tmp/after             │
                              │    sec/op    │    sec/op     vs base              │
LineBuffered/capacity:1-10      2.159µ ± ∞ ¹   2.227µ ± ∞ ¹       ~ (p=1.000 n=5)
LineBuffered/capacity:2-10      1.301µ ± ∞ ¹   1.212µ ± ∞ ¹       ~ (p=0.548 n=5)
LineBuffered/capacity:4-10      745.8n ± ∞ ¹   800.4n ± ∞ ¹       ~ (p=0.690 n=5)
LineBuffered/capacity:8-10      494.9n ± ∞ ¹   521.3n ± ∞ ¹       ~ (p=0.095 n=5)
LineBuffered/capacity:16-10     370.2n ± ∞ ¹   385.6n ± ∞ ¹       ~ (p=0.095 n=5)
LineBuffered/capacity:32-10     300.7n ± ∞ ¹   301.8n ± ∞ ¹       ~ (p=1.000 n=5)
LineBuffered/capacity:64-10     267.5n ± ∞ ¹   261.8n ± ∞ ¹       ~ (p=0.841 n=5)
LineBuffered/capacity:128-10    247.5n ± ∞ ¹   246.5n ± ∞ ¹       ~ (p=0.841 n=5)
LineBuffered/capacity:256-10    238.1n ± ∞ ¹   250.3n ± ∞ ¹  +5.12% (p=0.016 n=5)
LineBuffered/capacity:512-10    230.3n ± ∞ ¹   235.0n ± ∞ ¹       ~ (p=0.278 n=5)
LineBuffered/capacity:1024-10   225.1n ± ∞ ¹   238.1n ± ∞ ¹  +5.78% (p=0.016 n=5)
LineBuffered/capacity:2048-10   225.4n ± ∞ ¹   226.2n ± ∞ ¹       ~ (p=0.889 n=5)
geomean                         408.0n         415.9n        +1.95%
¹ need >= 6 samples for confidence interval at level 0.95

                              │ /tmp/before │             /tmp/after              │
                              │    B/op     │    B/op      vs base                │
LineBuffered/capacity:1-10      96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:2-10      96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:4-10      96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:8-10      96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:16-10     96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:32-10     96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:64-10     96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:128-10    96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:256-10    96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:512-10    96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:1024-10   96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:2048-10   96.00 ± ∞ ¹   96.00 ± ∞ ¹       ~ (p=1.000 n=5) ²
geomean                         96.00         96.00        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal

                              │ /tmp/before │             /tmp/after              │
                              │  allocs/op  │  allocs/op   vs base                │
LineBuffered/capacity:1-10      2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:2-10      2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:4-10      2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:8-10      2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:16-10     2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:32-10     2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:64-10     2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:128-10    2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:256-10    2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:512-10    2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:1024-10   2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
LineBuffered/capacity:2048-10   2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
geomean                         2.000         2.000        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
```

Basically seems the same to me.
I have no problem changing it, but I can't see much difference.

I'd rather keep it this way to reduce contention in case we start using `Read()` more in the future.

Reply:
Regardless the performance, I think it's unsafe to use a RWLock here. I can't understand how we can safely concurrently read from a `bytes.Buffer`.

Reply:
See: https://github.com/grafana/dskit/pull/339

Reply:
Thanks Marco, that makes sense